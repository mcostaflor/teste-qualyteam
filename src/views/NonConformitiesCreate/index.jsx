import React, { useState, useEffect } from 'react';
import { nonConformitiesApi, departmentsApi } from '../../services/api/routes';
import { useSnackbar } from 'notistack';
import dateFormat from 'dateformat';

import { TextField, FormGroup, FormControlLabel, Checkbox, Button, Card, CardContent } from '@material-ui/core';

export default function NonConformityCreate({ history }) {

    const [departments, setDepartments] = useState([]);
    const [description, setDescription] = useState('');
    const [ocurrenceDate, setOcurrenceDate] = useState(dateFormat(new Date(), 'yyyy-mm-dd'));

    const { enqueueSnackbar } = useSnackbar();

    useEffect(() => {
        async function fetchDepartments() {
            setDepartments(await departmentsApi.getAll());
        }

        fetchDepartments();
    }, []);

    const handleSubmit = async e => {
        e.preventDefault();

        let checkedDepartments = [];

        departments.map(item => {
            if (item.isChecked) {
                checkedDepartments.push(item.id);
            }
            return item;
        });

        if (ocurrenceDate === '') {
            enqueueSnackbar('Favor preencher o campo "Data da Ocorrência"', { variant: 'warning' });
            return;
        }
        if (description === '') {
            enqueueSnackbar('Favor preencher o campo "Descrição"', { variant: 'warning' });
            return;
        }
        if (checkedDepartments.length === 0) {
            enqueueSnackbar('Favor selecionar ao menos um departamento.', { variant: 'warning' })
            return;
        }

        const data = {
            description,
            'ocurrence-date': dateFormat(ocurrenceDate + 'T00:00:00', 'dd-mm-yyyy'),
            departments: checkedDepartments,
            'corrective-actions': []
        }

        try {
            const create = await nonConformitiesApi.createOne(data);
            history.push(`${create.id}`);
            enqueueSnackbar('Não conformidade criada com sucesso', { variant: 'success' });
        } catch (e) {
            enqueueSnackbar('Não criado.', { variant: 'error' })
        }

    }

    const handleDepartmentCheckbox = (e, id) => {
        let temp = [...departments];
        temp.find((item) => {
            return item.id === id;
        }).isChecked = e.target.checked;
        setDepartments(temp);
    }

    return (
        <>
            <Button onClick={() => { history.push('/nonconformities') }} variant="contained" color="primary">Voltar</Button>
            <div>
                <form id='new-nonconformity-form' onSubmit={handleSubmit} data-testid={'new-noncorformity-form'}>
                    <Card style={{ marginTop: 16 }}>
                        <CardContent>
                            <TextField
                                label={'Data da Ocorrência'}
                                type="date"
                                value={ocurrenceDate}
                                onChange={e => setOcurrenceDate(e.target.value)}
                                variant={'outlined'}
                                style={{
                                    marginTop: 18
                                }}
                                inputProps={{
                                    'data-testid': 'new-nonconformity-ocurrence-date'
                                }}
                            />
                            <TextField
                                label={'Descrição'}
                                multiline
                                rows={10}
                                value={description}
                                onChange={e => setDescription(e.target.value)}
                                variant={'outlined'}
                                fullWidth
                                style={{
                                    marginTop: 18
                                }}
                                inputProps={{
                                    'data-testid': 'new-nonconformity-description'
                                }}
                            />
                            <FormGroup label={'Departamento(s)'} row>
                                {departments.map((item, index) =>
                                    <FormControlLabel
                                        key={item.id}
                                        data-testid={`new-nonconformity-checkbox-${index}`}
                                        control={
                                            <Checkbox
                                                checked={item.isChecked || false}
                                                onChange={e => handleDepartmentCheckbox(e, item.id)}
                                            />
                                        }
                                        label={item.name}
                                    />
                                )}
                            </FormGroup>

                            <Button
                                type="submit"
                                variant="contained"
                                color="primary"
                                data-testid={'new-nonconformity-submit'}
                            >
                                Enviar
                            </Button>
                        </CardContent>
                    </Card>
                </form>
            </div>
        </>
    );
}