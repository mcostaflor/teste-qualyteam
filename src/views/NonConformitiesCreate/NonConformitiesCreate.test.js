import { render, cleanup, waitForElement, fireEvent, getByText } from '@testing-library/react'

import NonConformitiesCreate from './index';
import { renderApp } from '../../helpers/tests/renderApp';

afterEach(cleanup);

describe('Testing new non conformity submit view', () => {

    it("renders", async () => {
        const page = render(renderApp(NonConformitiesCreate));
        expect(page).toMatchSnapshot();
    });

    it("loads departments", async () => {
        const { getByTestId } = render(renderApp(NonConformitiesCreate));
        const checkbox = await waitForElement(
            () => getByTestId('new-nonconformity-checkbox-0')
        )
        expect(checkbox).toBeDefined();
    });

    it("sends the nonconformity data", async () => {

        const { getByTestId, getByText } = render(renderApp(NonConformitiesCreate));
        const description = getByTestId('new-nonconformity-description');
        const ocurrenceDate = getByTestId('new-nonconformity-ocurrence-date');
        const checkbox = await waitForElement(
            () =>
                getByTestId('new-nonconformity-checkbox-0')
        )

        expect(checkbox).toBeDefined();

        const submitButton = getByTestId('new-nonconformity-submit')

        fireEvent.change(description, {
            target: { value: 'Teste' }
        });

        fireEvent.change(ocurrenceDate, {
            target: { value: '2020-05-05' }
        });

        fireEvent.click(checkbox, {
            target: { checked: true }
        });

        fireEvent.click(submitButton);


        // history.push no componente retornando erro, cai no catch e dispara uma snackbar com a mensagem de não criado.
        const snackbar = await waitForElement(
            () => getByText('Não criado.')
        )

        expect(snackbar).toBeDefined();

    });

});