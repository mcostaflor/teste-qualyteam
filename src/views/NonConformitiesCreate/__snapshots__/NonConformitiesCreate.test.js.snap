// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing new non conformity submit view renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Voltar
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div>
        <form
          data-testid="new-noncorformity-form"
          id="new-nonconformity-form"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
            style="margin-top: 16px;"
          >
            <div
              class="MuiCardContent-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root"
                style="margin-top: 18px;"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                  data-shrink="true"
                >
                  Data da Ocorrência
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    data-testid="new-nonconformity-ocurrence-date"
                    type="date"
                    value="2020-03-08"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-121 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-123 PrivateNotchedOutline-legendNotched-124"
                    >
                      <span>
                        Data da Ocorrência
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                style="margin-top: 18px;"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                >
                  Descrição
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                    data-testid="new-nonconformity-description"
                    rows="10"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-121 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-123"
                    >
                      <span>
                        Descrição
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormGroup-root MuiFormGroup-row"
                label="Departamento(s)"
              />
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                data-testid="new-nonconformity-submit"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Enviar
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Voltar
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <form
        data-testid="new-noncorformity-form"
        id="new-nonconformity-form"
      >
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
          style="margin-top: 16px;"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root"
              style="margin-top: 18px;"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                data-shrink="true"
              >
                Data da Ocorrência
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  data-testid="new-nonconformity-ocurrence-date"
                  type="date"
                  value="2020-03-08"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-121 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-123 PrivateNotchedOutline-legendNotched-124"
                  >
                    <span>
                      Data da Ocorrência
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              style="margin-top: 18px;"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                data-shrink="false"
              >
                Descrição
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  data-testid="new-nonconformity-description"
                  rows="10"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-121 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-123"
                  >
                    <span>
                      Descrição
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormGroup-root MuiFormGroup-row"
              label="Departamento(s)"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              data-testid="new-nonconformity-submit"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Enviar
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
